### PR TITLE
feat: experimentation opportunities

### PR DESCRIPTION
**Pull request by ekremney**
This PR contains two new queries to be used in experimentation opportunity detection:
1. High inorganic traffic with high bounce rate
2. High organic traffic with low ctr

Tests and README will follow...

### CODE REVIEW

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    pageViews: total,
+    samples: total, // todo: get the actual number of samples
+    metrics: [], // none needed
+  };
+}
+
+function hasHighInorganicTraffic(traffic) {
+  const { url, paid, total } = traffic;
+  const isHomapage = new URL(url).pathname === '/';
+  const threshold = isHomapage ? 0.5 : 0.8;
```
Comment (ramboz) [inline packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js:34]: Would there be a need for those to be site-specific and passed as configuration or query parameters somehow? or can we keep these static values for all

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  return paid / total > threshold;
+}
+
+function hasHighBounceRate(ctr) {
+  return ctr < 0.5;
```
Comment (ramboz) [inline packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js:39]: Same question here

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import trafficAcquisition from '../traffic-acquisition.js';
+import { getCTRByUrl, getSiteAvgCTR } from '../../common/aggregateFns.js';
+
+const DAILY_EARNED_THRESHOLD = 5000;
```
Comment (ramboz) [inline packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js:16]: Same question

**Review by ramboz — APPROVED** [review]

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js (hunk 1 above)
Comment (rpapani) [reply packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js:34]: better to make these as configurable as env variables

File: packages/spacecat-shared-rum-api-client/src/common/aggregateFns.js [hunk 4]
```diff
@@ -23,6 +23,60 @@ function pageviewsByUrl(bundles) {
...
+    .reduce((acc, [url, { sessionsWithClick, totalPageviews }]) => {
+      acc[url] = (sessionsWithClick / totalPageviews);
+      return acc;
+    }, {});
+}
+
+/**
+ * Calculates the Click-Through Rate (CTR) average for the entire site.
+ * CTR is defined as the total number of sessions with at least one click event
+ * divided by the total number of pageviews for each URL.
```
Comment (rpapani) [inline packages/spacecat-shared-rum-api-client/src/common/aggregateFns.js:58]: this is entire site CTR, not the CTR per URL right? may be, can we remove `for each URL` in that case?

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+  return {
+    type: 'high-inorganic-high-bounce-rate',
+    page: url,
+    screenshot: '',
+    trackedPageKPIName: 'Bounce Rate',
+    trackedPageKPIValue: bounceRate,
+    pageViews: total,
+    samples: total, // todo: get the actual number of samples
+    metrics: [], // none needed
```
Comment (rpapani) [inline packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js:27]: Can we provide, how much of inorganic traffic is coming and how much of them are engaging, in the metrics?

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  };
+}
+
+function hasHighOrganicTraffic(interval, traffic) {
+  const { earned } = traffic;
+  return earned > DAILY_EARNED_THRESHOLD * interval;
+}
+
+function hasLowerCTR(ctr, siteAvgCTR) {
+  return ctr < 0.95 * siteAvgCTR;
```
Comment (rpapani) [inline packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js:39]: can we extract this to `const`?

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+  return {
+    type: 'high-organic-low-ctr',
+    page: url,
+    screenshot: '',
+    trackedPageKPIName: 'Click Through Rate',
+    trackedPageKPIValue: ctr,
+    pageViews: total,
+    samples: total, // todo: get the actual number of samples
+    metrics: [], // none needed
```
Comment (rpapani) [inline packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js:29]: may be its better to add the siteAverageCTR and page CTR in the metrics

**Comment by rpapani** [comment]
@ekremney when I run this on adobe.com, I'm getting lot of opportunities which have very few page views, I think we should introduce some page views threshold to remove this noise.
```
{
  "type": "high-inorganic-high-bounce-rate",
  "page": "https://www.adobe.com/creativecloud/business/teams/resources/how-to/envelope-design.html",
  "screenshot": "",
  "trackedPageKPIName": "Bounce Rate",
  "trackedPageKPIValue": 1,
  "pageViews": 100,
  "samples": 100,
  "metrics": []
}
```

File: packages/spacecat-shared-rum-api-client/src/common/aggregateFns.js (hunk 4 above)
Comment (ekremney) [reply packages/spacecat-shared-rum-api-client/src/common/aggregateFns.js:58]: copy/pasta error 👍

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js (hunk 6 above)
Comment (ekremney) [reply packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js:39]: done

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js (hunk 1 above)
Comment (ekremney) [reply packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js]: I extracted those values to constant.

For now, I'd say let's start with constants. Later-on we can make it configurable per site if needed. This should be data-driven, I really don't know what values best for each customer at the moment.

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js (hunk 2 above)
Comment (ekremney) [reply packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-inorganic-high-bounce-rate.js]: same answer here https://github.com/adobe/spacecat-shared/pull/351#discussion_r1741742304

File: packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js (hunk 7 above)
Comment (ekremney) [reply packages/spacecat-shared-rum-api-client/src/functions/opportunities/high-organic-low-ctr.js]: done

**Comment by ekremney** [comment]
@ramboz @rpapani gentlemen, I did the following improvements as per your reviews:
1. the threshold values are extracted to constants. I don't know about making them configurable per site, this is something we should observe and iterate in the future
2. I added extra information such as traffic sources, ctr etc. to the metrics array 
3. I set the daily pageview threshold to 1000 in order for pages to qualify for "opportunities"
4. other various changes...

**Pull request merged**